crates/bindings/src/service.rs

```diff
@@ -0,0 +1,230 @@
+/// Service is a trait implemented by services which may be called from Go.
+pub trait Service {
+    /// Create a new instance of the Service.
+    fn create() -> Self;
+    /// Invoke with the given op-code & |data| payload.
+    /// It extends |arena| with any returned []byte data, and pushes output messages onto |out|.
+    fn invoke(&mut self, code: u32, data: &[u8], arena: &mut Vec<u8>, out: &mut Vec<Out>);
+}
+
+/// Input frame produced from CGO, which is a single service invocation.
+/// 16 bytes, or 1/4 of a typical cache line.
+#[repr(C)]
+pub struct In1 {
+    data_ptr: *const u8,
+    data_len: u32,
+    code: u32,
+}
+
+/// Four invocations, composed into one struct.
+/// 64 bytes, or one typical cache line.
+#[repr(C)]
+pub struct In4 {
+    in0: In1,
+    in1: In1,
+    in2: In1,
+    in3: In1,
+}
+
+/// Sixteen invocations, composed into one struct.
+/// 256 bytes, or four typical cache lines.
+#[repr(C)]
+pub struct In16 {
+    in00: In1,
+    in01: In1,
+    in02: In1,
+    in03: In1,
+
+    in04: In1,
+    in05: In1,
+    in06: In1,
+    in07: In1,
+
+    in08: In1,
+    in09: In1,
+    in10: In1,
+    in11: In1,
+
+    in12: In1,
+    in13: In1,
+    in14: In1,
+    in15: In1,
+}
+
+/// Output
+#[repr(C)]
+pub struct Out {
+    pub code: u32,
+    pub begin: u32,
+    pub end: u32,
+}
+
+/// Opaque pointer for a Service instance.
+#[repr(C)]
+pub struct ServiceImpl {
+    _private: [u8; 0],
+}
+
+/// Channel is shared between CGO and Rust, and holds details
+/// about the language interconnect.
+#[repr(C)]
+pub struct Channel {
+    // Output memory arena, exposed to CGO.
+    arena_ptr: *mut u8,
+    arena_len: usize,
+    arena_cap: usize,
+
+    // Output frame codes & arena offsets, exposed to CGO.
+    out_ptr: *mut Out,
+    out_len: usize,
+    out_cap: usize,
+
+    // Opaque service pointer.
+    svc_impl: *mut ServiceImpl,
+}
+
+/// Create a new Service instance, wrapped in an owning Channel.
+/// This is intended to be monomorphized by each Service implementation,
+/// and exposed via cbindgen.  See the UpperCase service for an example.
+#[inline]
```

Comment:
What's the purpose of the `#[inline]`? I wouldn't expect calls to these to ever be inlined in practice since it's called from go.

Reply:
These aren't called from go, they're called from monomorphized one-liners for each specific service.